### PR TITLE
[LSP] Cache parsed xml snippets for razor

### PR DESCRIPTION
**Pull request by dibarbet**
A followup change to the original PR.   We can easily cache the `SnippetPart`s we get from parsing the snippet XML so that we don't have to re-parse them from XML on each request.

### CODE REVIEW

File: src/Features/LanguageServer/Protocol/Handler/InlineCompletions/XmlSnippetParser.cs [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+        }
+
+        var expansion = new ExpansionTemplate(matchingSnippet);
+
+        // Parse the snippet XML into snippet parts we can cache.
+        var parsedSnippet = expansion.Parse();
+        return parsedSnippet;
+    }
+
+    private static CodeSnippet RetrieveSnippetXmlFromFile(SnippetInfo snippetInfo)
```
Comment (dibarbet) [inline src/Features/LanguageServer/Protocol/Handler/InlineCompletions/XmlSnippetParser.cs:87]: Moved from
https://github.com/dotnet/roslyn/pull/59605/files#diff-161082fbc336f297b8314bceb1e12ca75fae05cf30280d7b3dfe3fe35f91f70aL270

File: src/Features/LanguageServer/Protocol/Handler/InlineCompletions/XmlSnippetParser.cs [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+            // Add the snippet to the cache regardless of if we succeeded in parsing it.
+            // We're not likely to succeed in parsing on a second try if we failed initially, so we cache it to avoid repeatedly failing.
+            _parsedSnippetsCache.TryAdd(matchingSnippetInfo.Title, parsedSnippet);
+        }
+    }
+
+    private static ParsedXmlSnippet GetAndParseSnippetFromFile(SnippetInfo snippetInfo)
+    {
+        // Read the XML file to get the snippet and snippet metadata.
+        var matchingSnippet = RetrieveSnippetXmlFromFile(snippetInfo);
```
Comment (dibarbet) [inline src/Features/LanguageServer/Protocol/Handler/InlineCompletions/XmlSnippetParser.cs:71]: moved from https://github.com/dotnet/roslyn/pull/59605/files#diff-161082fbc336f297b8314bceb1e12ca75fae05cf30280d7b3dfe3fe35f91f70aL91

File: src/Features/LanguageServer/Protocol/Handler/InlineCompletions/InlineCompletionsHandler.cs [hunk 3]
```diff
@@ -22,6 +20,7 @@
 using Microsoft.CodeAnalysis.Text;
 using Microsoft.VisualStudio.LanguageServer.Protocol;
 using Roslyn.Utilities;
+using static Microsoft.CodeAnalysis.LanguageServer.Handler.InlineCompletions.XmlSnippetParser;
```
Comment (davidwengier) [inline src/Features/LanguageServer/Protocol/Handler/InlineCompletions/InlineCompletionsHandler.cs:23]: Seems unnecessary

File: src/Features/LanguageServer/Protocol/Handler/InlineCompletions/XmlSnippetParser.cs [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+        {
+            context.TraceInformation($"Reading snippet for {matchingSnippetInfo.Title} with path {matchingSnippetInfo.Path}");
+            parsedSnippet = GetAndParseSnippetFromFile(matchingSnippetInfo);
+            return parsedSnippet;
+        }
+        catch (Exception ex) when (FatalError.ReportAndCatch(ex, ErrorSeverity.General))
+        {
+            context.TraceError($"Got exception parsing xml snippet {matchingSnippetInfo.Title} from file {matchingSnippetInfo.Path}");
+            context.TraceException(ex);
+            return null;
```
Comment (davidwengier) [inline src/Features/LanguageServer/Protocol/Handler/InlineCompletions/XmlSnippetParser.cs:58]: Nit: Don't need this return, or the one above, or the finally.. can just return `parsedSnippet` at the end. Would make it a little easier to follow

**Review by davidwengier — APPROVED** [review]

File: src/Features/LanguageServer/Protocol/Handler/InlineCompletions/XmlSnippetParser.cs (hunk 4 above)
Comment (dibarbet) [reply src/Features/LanguageServer/Protocol/Handler/InlineCompletions/XmlSnippetParser.cs]: 🤦fixed!

File: src/Features/LanguageServer/Protocol/Handler/InlineCompletions/InlineCompletionsHandler.cs (hunk 3 above)
Comment (dibarbet) [reply src/Features/LanguageServer/Protocol/Handler/InlineCompletions/InlineCompletionsHandler.cs:23]: a couple of types from the parser are used here

**Review by NTaylorMullen — APPROVED** [review]
Dopeness and great comments! :shipit:

**Pull request merged**